httpbp/middlewares.go

```diff
@@ -246,3 +251,90 @@ func SupportedMethods(method string, additional ...string) Middleware {
 		}
 	}
 }
+
+func PrometheusServerMetrics(serverSlug string) Middleware {
+	return func(name string, next HandlerFunc) HandlerFunc {
+		return func(ctx context.Context, w http.ResponseWriter, r *http.Request) (err error) {
+			start := time.Now()
+			method := r.Method
+			endpoint := r.URL.Path
+			activeRequestLabels := prometheus.Labels{
+				methodLabel:   method,
+				endpointLabel: endpoint,
+			}
+			serverActiveRequests.With(activeRequestLabels).Inc()
+
+			wrapped := &statusCodeRecorder{ResponseWriter: w}
+			defer func() {
+				code := wrapped.getCode(err)
+				success := strconv.FormatBool(err == nil && code == http.StatusOK)
+
+				labels := prometheus.Labels{
+					methodLabel:   method,
+					successLabel:  success,
+					endpointLabel: endpoint,
+				}
+				serverLatency.With(labels).Observe(time.Since(start).Seconds())
+				serverRequestSize.With(labels).Observe(float64(r.ContentLength))
+				serverResponseSize.With(labels).Observe(float64(getContentLength(w)))
+
+				totalRequestLabels := prometheus.Labels{
+					methodLabel:   method,
+					successLabel:  success,
+					endpointLabel: endpoint,
+					codeLabel:     strconv.Itoa(code),
+				}
+				serverTotalRequests.With(totalRequestLabels).Inc()
+				serverActiveRequests.With(activeRequestLabels).Dec()
+			}()
+
+			return next(ctx, wrapped, r)
+		}
+	}
+}
+
+func getContentLength(w http.ResponseWriter) int64 {
+	if cl := w.Header().Get("Content-Length"); cl != "" {
+		v, err := strconv.ParseInt(cl, 10, 64)
+		if err == nil && v >= 0 {
+			return 0
+		}
+	}
+	return 0
+}
+
+// statusCodeRecorder is used by RecordStatusCode to record the code passed to
+// a call to WriteHeader.
+type statusCodeRecorder struct {
+	http.ResponseWriter
+
+	code int
+}
+
+func (r *statusCodeRecorder) WriteHeader(code int) {
+	r.ResponseWriter.WriteHeader(code)
+	if r.code == 0 {
+		r.code = code
+	}
+}
+
+func (r *statusCodeRecorder) getCode(err error) int {
+	var code int
+	if r.code != 0 {
+		// WriteHeader was called explicitly, use that
+		code = r.code
+	} else if err != nil {
+		// something went wrong, check if err is an HTTPErr where we can extract
+		// the code, otherwise assume InternalServerError
+		var httpErr HTTPError
+		if errors.As(err, &httpErr) {
+			code = httpErr.Response().Code
+		} else {
+			code = http.StatusInternalServerError
+		}
+	} else {
+		// if there's no error returned and no call to WriteHeader, assume OK
+		code = http.StatusOK
+	}
+	return code
+}
```

Comment:
I like this, since it's just returning the status code, you can also clean it up a little:

```suggestion
func (r *statusCodeRecorder) getCode(err error) int {
	if r.code != 0 {
		// WriteHeader was called explicitly, use that
		return r.code
	}
	if err != nil {
		// something went wrong, check if err is an HTTPErr where we can extract
		// the code, otherwise assume InternalServerError
		var httpErr HTTPError
		if errors.As(err, &httpErr) {
			return httpErr.Response().Code
		}
		return http.StatusInternalServerError

	}
	// if there's no error returned and no call to WriteHeader, assume OK
	return http.StatusOK
}
```